src/libs/antares/study/parts/renewable/container.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+** Copyright 2007-2018 RTE
+** Authors: Antares_Simulator Team
+**
+** This file is part of Antares_Simulator.
+**
+** Antares_Simulator is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** There are special exceptions to the terms and conditions of the
+** license as they are applied to this software. View the full text of
+** the exceptions in file COPYING.txt in the directory of this software
+** distribution
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with Antares_Simulator. If not, see <http://www.gnu.org/licenses/>.
+**
+** SPDX-License-Identifier: licenceRef-GPL3_WITH_RTE-Exceptions
+*/
+
+#include <yuni/yuni.h>
+#include "../../study.h"
+#include "../../memory-usage.h"
+#include "container.h"
+#include "../../../logs.h"
+
+using namespace Yuni;
+using namespace Antares;
+
+namespace Antares
+{
+namespace Data
+{
+PartRenewable::PartRenewable() : clusters(nullptr), clusterCount((uint)-1)
+{
+}
+
+bool PartRenewable::invalidate(bool reload) const
+{
+    bool ret = true;
+    ret = list.invalidate(reload) && ret;
+    return ret;
+}
+
+void PartRenewable::markAsModified() const
+{
+    list.markAsModified();
+}
+
+void PartRenewable::estimateMemoryUsage(StudyMemoryUsage& u) const
+{
+    u.requiredMemoryForInput += sizeof(PartRenewable);
+    list.estimateMemoryUsage(u);
+}
+
+PartRenewable::~PartRenewable()
+{
+    if (clusterCount)
+        delete[] clusters;
+}
+
+void PartRenewable::prepareAreaWideIndexes()
+{
+    // Copy the list with all renewable clusters
+    // And init the areaWideIndex (unique index for a given area)
+    clusterCount = list.size();
+    delete[] clusters;
+    if (!clusterCount)
+    {
+        clusters = nullptr;
+        return;
+    }
+
+    typedef RenewableCluster* RenewableClusterPointer;
+    clusters = new RenewableClusterPointer[clusterCount];
+
+    auto end = list.end();
+    uint idx = 0;
+    for (auto i = list.begin(); i != end; ++i)
+    {
+        RenewableCluster* t = i->second;
+        t->areaWideIndex = idx;
+        clusters[idx] = t;
+        ++idx;
+    }
+}
+
+uint PartRenewable::removeDisabledClusters()
```

Comment:
why is this function so complicated, with 2 loops (**do** and **for**) intertwined ?

Reply:
These two loops are used to remove every disabled cluster from the list.

Iterator `list.end()` can change and become invalid if `list.remove(i);` is called. In this case, we get out of the `for` (with the `break` instruction) loop to reset this iterator.

If `list.remove(i)` does not invalidate `list.end()`, the two loops are not needed. But this assumption is unclear.

Reply:
Re-written in pair programming.